### ISSUE TITLE
Not working correctly on Android 10

### ISSUE THREAD

**Issue by IzzySoft** [Limizin/GpsIcon#3]
Having reached Android 10, I'm back at an old issue, just different: the icon is shown, but permanently. It always shows 0/0 satellites, even with a fix. That's on a Fairphone 2 running LineageOS. My brother-in-law uses a Fairphone 3 with stock Android 10 and does not have that issue.

I've checked logcat but found no clue there. The only entries directly mentioning the app are

```
W Role    : dev.ldev.gpsicon not qualified for android.app.role.HOME due to missing RequiredComponent{mIntentFilterData=IntentFilterData{mAction='android.intent.action.MAIN', mCategories='[android.intent.category.HOME]', mDataScheme='null', mDataType='null'}, mPermission='null', mMetaData=[]}
W Role    : dev.ldev.gpsicon not qualified for android.app.role.DIALER due to missing RequiredComponent{mIntentFilterData=IntentFilterData{mAction='android.intent.action.DIAL', mCategories='[]', mDataScheme='null', mDataType='null'}, mPermission='null', mMetaData=[]}
W Role    : dev.ldev.gpsicon not qualified for android.app.role.SMS due to missing RequiredComponent{mIntentFilterData=IntentFilterData{mAction='android.provider.Telephony.SMS_DELIVER', mCategories='[]', mDataScheme='null', mDataType='null'}, mPermission='android.permission.BROADCAST_SMS', mMetaData=[]}
```

which do not seem related. Just trying it again now: SatStat sees 3/32, GpsIcon 0/0 – and this time something in the logs that could be related:

```
E Layer   : [Surface(name=AppWindowToken{c3f5008 token=Token{5b191ab ActivityRecord{ad117fa u0 dev.ldev.gpsicon/.ui.MainActivity t132}}})/@0xfa89326 - animation-leash#0] No local sync point found
E Layer   : [Surface(name=AppWindowToken{c3f5008 token=Token{5b191ab ActivityRecord{ad117fa u0 dev.ldev.gpsicon/.ui.MainActivity t132}}})/@0xfa89326 - animation-leash#0] No local sync point found
E Layer   : [Surface(name=AppWindowToken{c3f5008 token=Token{5b191ab ActivityRecord{ad117fa u0 dev.ldev.gpsicon/.ui.MainActivity t132}}})/@0xfa89326 - animation-leash#0] No local sync point found
E Layer   : [Surface(name=AppWindowToken{c3f5008 token=Token{5b191ab ActivityRecord{ad117fa u0 dev.ldev.gpsicon/.ui.MainActivity t132}}})/@0xfa89326 - animation-leash#0] No local sync point found
E Layer   : [Surface(name=AppWindowToken{791f978 token=Token{c83adb ActivityRecord{bc0e7ea u0 dev.ldev.gpsicon/.ui.MainActivity t133}}})/@0x2db1f9a - animation-leash#0] No local sync point found
E Layer   : [Surface(name=AppWindowToken{f6b3254 token=Token{fa99da7 ActivityRecord{dca2666 u0 dev.ldev.gpsicon/.ui.MainActivity t134}}})/@0x366ab4 - animation-leash#0] No local sync point found
E Layer   : [Surface(name=AppWindowToken{f6b3254 token=Token{fa99da7 ActivityRecord{dca2666 u0 dev.ldev.gpsicon/.ui.MainActivity t134}}})/@0x366ab4 - animation-leash#0] No local sync point found
```

Though this could also come from my toggling between digital and blinking indicator.

Do you have any idea what this could be, and how to fix it? I hope you're still around…

**Comment by Limizin** [comment]
>No local sync point found

This is something LineageOS internal, but not a reason 

The reason is that 
     
        _locationManager.addGpsStatusListener(this);
        ...
         void onGpsStatusChanged(int event) {

the method `onGpsStatusChanged` is not called. I do not know why.  Maybe [because ](https://developer.android.com/reference/android/location/LocationManager#addGpsStatusListener(android.location.GpsStatus.Listener))

      This method was deprecated in API level 24.
      use registerGnssStatusCallback(android.location.GnssStatus.Callback) instead. 
      No longer supported in apps targeting S and above. 

But it's not enough just up api level to 24 and use `registerGnssStatusCallback`. The system will start killing the service (an innovation in oreo), so the application has api level 23 to get around this limitation 

I definitely can't do it. I'm not an android developer. I wrote this application by googling and did not write it for android anymore.
 
And I do not plan to, since I still use android 5 (I can not find a compact smartphone with large battery and IPS)

**Comment by IzzySoft** [comment]
Thanks for your reply, @Limizin – tbh, I wasn't even sure if you're still around, so I'm happy you are! I've [tooted a callout for help](https://mastodon.technology/@IzzyOnDroid/106524772892796624) which already got quite a few favs and boosts, so I'm hoping someone will take a look soon and provide a PR.

What puzzles me is that it's not working on my FP2 running Android 10 with LOS, while it works fine on a FP3 running stock Android 10 (the one of my brother-in-law).

> No longer supported in apps targeting S and above.

To me that sounds it should still be available to apps targeting R and below.

> I still use android 5

Ouch. The only LP device I still use is my 7" tablet for eBook reading. Completely offline of course.

> I can not find a compact smartphone with large battery and IPS

Always hard to find exact matches for one's needs. Define compact and large. There are some sites allowing you to filter for such things.

**Comment by Limizin** [comment]
>There are some sites allowing you to filter for such things.

there are sites, no compacts. a couple of years ago I did not change my smartphone, and now there is nothing - only phablets 
>To me that sounds it should still be available to apps targeting R and below.

Right. It should, but it doesn't work.  Rather it works, [but in its own way ](https://developer.android.com/about/versions/oreo/background-location-limits)

> In an effort to reduce power consumption, Android 8.0 (API level 26) limits how frequently an app can retrieve the user's current location while the app is running in the background. Under these conditions, apps can receive location updates only a few times each hour.

So minimal way to fix it. In this case, the icon will always be visible. 

- Use `_context.startForegroundService(new Intent(_context, GpsObserveService.class));` instead of `_context.startService(new Intent(_context, GpsObserveService.class));`
- add permission `<uses-permission android:name="android.permission.FOREGROUND_SERVICE" />`
- "[If your app targets Android 10 (API level 29) or higher, also check for the ACCESS_BACKGROUND_LOCATION permission. ](https://developer.android.com/training/location/background)"
- fix [notification for oreo+ ](https://stackoverflow.com/questions/44658923/android-foreground-service-notification-not-showing)
- Make sure the service does not crash

**Comment by IzzySoft** [comment]
> there are sites, no compacts. a couple of years ago I did not change my smartphone, and now there is nothing - only phablets

* [Handyfinder](https://www.areamobile.de/Handyfinder/)
* [another one](https://geizhals.de/?o=39)

For example, [213 Android devices up to 4.7"](https://geizhals.de/?cat=umtsover&xf=148_Android%7E2392_4.7) (you can shrink further, and choose a min battery capacity as well).

I'm lost with the API details, though (I'm not and Android dev) – but yes, I've noticed that much more apps require a permanent notification icon nowadays, which I often find quite annoying. In case of GpsIcon, this could be limited a bit by have it only start when location is turned on, and stop the service when location service is disabled. Not sure if there's a broadcast for "location is in use" – if there's such, that would help keeping the current behavior.

**Comment by Limizin** [comment]
try this. It is debug build.
[app-debug.zip](https://github.com/Limizin/GpsIcon/files/6766474/app-debug.zip)

I hate google and android 8+ 
(+) mean 'realized'...may be
1 oreo+ requires notification channel (+)
2 foreground service requires special notification otherwise after 5 seconds the service kills (+)
3 location changed now register in runtime (it's crazy) (+)
4 restarting service after oom. I have no idea how to do it (-)
5 after installation, you cannot touch GUI until the permissions are granted, otherwise the permission request icon will disappear (-)

MHO only item 4 is not enough for normal personal use

**Comment by IzzySoft** [comment]
Thanks a lot! Sitting in my room certainly isn't the best place to test, but still:

* granted the location permission via Android settings directly (wasn't offered the "always" option)
* started the app to set it to digital ("always show" can no longer be unticked it seems; you already indicated something along those lines)
* started some app to request location (OsmAnd)
* Eagerly open the notifications and indeed now see "0/32" – no longer "0/0", so it definitely did something!
* cross-checking with OsmAnd: numbers differ, OsmAnd toggles between 0/26 and 0/27 while GpsIcon sticks to 0/32 (the "0" is clear for all of them: no reception in this room with the blinds down)
* switching OsmAnd off is detected fine: GpsIcon immediately stops blinking, stating "GPS inactive" (that didn't work before either, so it indicates something is working now)

Will test a bit more tomorrow, time to go to bed now (just having left the shower I don't want to open the blinds and go outside now). Will check my mail before that, just in case you've got another build :stuck_out_tongue_winking_eye:

**Comment by Limizin** [comment]
It's unstable. If you turn on-off-on-off geolocation, it may stop working

**Comment by IzzySoft** [comment]
No need to turn off/on. I just tried today outside: back at 0/0 while OsmAnd showed 4/32. But it still detects whether GPS is in use (which it didn't before).

**Comment by Limizin** [comment]
[app-debug2.zip](https://github.com/Limizin/GpsIcon/files/6768392/app-debug2.zip)
It is debug build - it writes debug log to logcat
also rebuilt under SDK 26 otherwise it just stalls

**Comment by IzzySoft** [comment]
Wow! This one seems to work fine. Just tested with SatStat, GpsIcon was always half a second faster but apart from that in sync with what SatStat showed. Will keep an eye on it what it's doing "in the long run". Will report back then. If you have specific steps I should try, just let me know.

**Comment by Limizin** [comment]
The background service, due to the START_STICKY flag, was restarted by the system itself if it was killed. Not everything is clear with foreground service.

It seems that closing the application in "recent" does not affect anything (and it was expected that it will kill the service) - this should be tested

It also seems that app need to somehow restart the service self if it was killed by the by oom - but I don't know how to get the oom for the test. I can only write code about which I will never know if it works 

In general, you need to try to kill him in every possible way. 
Also fast turn off/on/off/on/off in different modes of application operation 

upd: if close app from 'recent' when location is off - the application will not come to life by itself. need to look for a way to be reborn. Previously, the "location chnaged" subscription did not require a running application, but now it does.

**Comment by Limizin** [comment]
[GpsIcon1.4.6(Q).zip](https://github.com/Limizin/GpsIcon/files/6796434/GpsIcon1.4.6.Q.zip)

Signed release. This is all that I could. android O+ requires the service to be running all the time in order to catch location on / off events. But he also requires that there be a notification. And if I don't want to see the icon unnecessarily, then I have to keep the application running. How do I prevent killing the application? No idea. So all I can do is remove the app from recent and add some code to resurrect the app after killing oom which I can't even test

**Comment by IzzySoft** [comment]
Yeah, that O+ requirement can be quite a pain in the … back – especially on smaller screens and if you have multiple apps having to meet it (here e.g. Tasker, AFWall+ and now GpsIcon). It fills the notification area unnecessarily. Unfortunately not much you can do about that with Android enforcing it (for whatever reasons).

I'm not an Android dev, so I've got no idea for alternative approaches of keeping an app from being killed by Doze or OOM. For the latter I vaguely remember one could play with values like `oom_adj` – for an app as small as yours that should be fine. As for doze, you could have it opt out from battery optimization.

For the signed release: Will you push the changes to the repo and make it a new version? 1.4.6 has been around for a while already :wink:

Thanks a lot for all your efforts! A pity that my CallOut on Mastodon did not bring any help here, though it was boosted and liked quite a bit. Probably gained you a few more users…

**Comment by Limizin** [comment]
I have pushed, but I use mercurial and I don’t know and I don’t want to know git - and instead of new branch it pushed on top... However, this is a failure  - the android eventually unloads the application and I do not know how to prevent this.

**Comment by IzzySoft** [comment]
Push arrived. Maybe make [this](https://github.com/Limizin/GpsIcon/blob/master/app/build.gradle#L10) `versionCode: 11` and `versionName: '1.4.7'` (or anything of your choice)?

> the android eventually unloads the application

I still keep watching it on my FP2 running Android 10. So far, it works reliably. But I didn't try to cause it any trouble, just was using it as I always do. And am happy that it works, thanks a lot!

> and I do not know how to prevent this.

I'll add [another shout](https://mastodon.technology/@IzzyOnDroid/106563629001118844) at Mastodon. This is a question so generic that most Android devs should have asked it one point or the other.

When adding the new version to my repo, I'll put a note to stick with 1.4.6 unless it fails. Btw, did you decide for a license? If you don't care, there's always WTFPL and Unlicense :laughing:

**Comment by Limizin** [comment]
It's just that you're in luck. In my case after 1-2 hours of sleep, the system unloads the application, and this is despite the flag in the settings "turn off battery optimization for this application". Suitable only for the lucky ones. Therefore, there is no sense in making a release 

> If you don't care, there's always WTFPL and Unlicense 

Any license will suitable for me. After all, in fact, the application is abandoned

**Comment by IzzySoft** [comment]
> and this is despite the flag in the settings "turn off battery optimization for this application"

So have you been to Android settings and exempted it from battery optimization? AFAIK that flag just allows that step to be performed, but the user must have the final say to use it. But this might require Android 6+ (Doze was introduced with 6 IIRC). Not sure how 5 handles that.

> After all, in fact, the application is abandoned

Ugh… Then a special thanks for maintaining an abandoned app I consider essential :smiley: I cannot simply declare any license in my repo if you didn't declare it for the app, btw.

**Comment by Limizin** [comment]
>Not sure how 5 handles that.

This is  tablet for test with android 10 (almost AOSP) 

>if you didn't declare it for the app, btw.

choose a license and tell me what to do

**Comment by IzzySoft** [comment]
>  tablet for test with android 10

Ah, OK – there you should find exclusion from battery saving in the system settings.

> choose a license and tell me what to do

[Here's the howto](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository) for adding a license. If you want to make the code "public domain", [The Unlicense](https://spdx.org/licenses/Unlicense.html) would probably be the best choice. Otherwise you could pick any [A]GPL-2/3 from [this list](https://spdx.org/licenses/) (they are more restrictive); [MIT](https://spdx.org/licenses/MIT.html) and [Apache-2](https://spdx.org/licenses/Apache-2.0.html) are two of the most used permissive licenses, BSD-2 & 3 are also frequent. Not much ideas about the rest.

So basically, from most permissive to most restrictive I'd narrow down to: Unlicense › BSD/MIT/Apache › GPL3 › GPL2/AGPL. Shortest to read: Unlicense, BSD, MIT. Reading the GPLs will take a while :rofl:

**Comment by Limizin** [comment]
I chose Unlicense 

In any case, there is no way to keep an application running without a foreground service. Well, maybe use JobScheduler to periodically try to restart, but this is unreliable

**Comment by IzzySoft** [comment]
> I chose Unlicense

Thanks! Thought so. Updated metadata on my end correspondingly.

> there is no way to keep an application running without a foreground service.

Isn't that what you're currently doing? At least on my FP2, it is still running (from the first start).

**Comment by Limizin** [comment]
>Isn't that what you're currently doing? 

Not. I start the foreground service only when the location is turned on - and the system will not touch this service.
But to enable the service, app need to subscribe to the "enable location" event - and according to the new rules, only the active running application can do this. If the application is unloaded, then it will stop receiving this notification. And so it does not work to keep it in memory - it is not a service, there is no icon, and therefore the system kills it 

To keep an always running service with a notification just to use gps sometimes - I don't want to 

I just removed the application from the "recent" ones so that the user does not interfere and he does not close it, as well as a possible restart if there is not enough memory, but this does not affect the unloading in any way - I have 6GB RAM on device with android 10 and 50% are always free

**Comment by IzzySoft** [comment]
> I start the foreground service only when the location is turned on

Ah, and because I never turned location entirely off, GpsIcon kept running. Yeah, that explains.

> according to the new rules

which are only to protect us… I wish we were trusted with the decision which services we'd like running and which not – instead of having others decide what's in our best interest. Seems like what's really in our own interest is colliding with "theirs"…

So OK, one way to work around this is to have an app like Tasker start GpsIcon whenever location is turned on. Not elegant, but a work-around. If you happen to have such an app running, that is…

**Comment by Limizin** [comment]
I wondered how the tasker kept itself running. I'm not sure, but I understood this - notification hannels appeared in android oreo + and one can configure each channel separately. Tasker creates a special notification and asks to hide it.

This means that I need to create an eternal service with a notification, which will observe location on/off and a separate notification channel for the satellite counter.

But I can’t modify code for all platforms. So, if I have time, I will try to fork repo and rebuild project for foreground service (for android 10+)

**Comment by IzzySoft** [comment]
That sounds fantastic! And take your time: the "temporary version" seems to do the job for me on 10 – and as nobody else complained loudly, pressure seems to be low (though I got a bunch of likes for my Mastodon toot, so there seems to be interest).

When rewriting, you might consider one of the following options:

* choose a different package name
* set minSdk to Android 9 or 10

Both could of course be combined. Idea behind that: as up to Oreo the current version works fine, and functionality remains the same, this avoids "accidental updates" on Oreo and below. Also, this allows you to concentrate on select platforms (I took "I can’t modify code for all platforms" as referring to Android versions), e.g. Pie/Q and up, not having to take care for Oreo and below.

> I will try to fork repo and rebuild project for foreground service (for android 10+)

Yupp, sounds good. Please make sure to set minSdk accordingly then :wink:

Thanks a lot for taking care – and not giving up!

**Comment by IzzySoft** [comment]
Just as a heads-up: that "temporary version" still works wonderfully on Android 11. Maybe it should be tagged and added as a new version to my repo, too?

**Comment by Limizin** [comment]
>still works wonderfully on Android 11

But it's not. It's just that you're lucky that the system doesn't kill the process. Therefore, it is correct to say "may work"

**Comment by IzzySoft** [comment]
OK, that might be. Works for about a year now without any issues, first on Android 10 (LOS 17) then 11 (LOS 18) on a FP2. End of the month I get my ShiftPhone, we'll see how well it works there then. Thanks anyway – and I'll not put it public unless you say :wink:

**Comment by Limizin** [comment]
You can do whatever you want. That is, everything. For me, this repository is abandoned. In my current smartphone, gps works great and I no longer use this program. If suddenly I have time (highly unlikely) and the desire to write a version for 10+, then this will be a new repository from scratch.

**Comment by IzzySoft** [comment]
Fair enough, thanks! In that as you wrote unlikely event, it'd be great if you remember me and let me know. Thanks a lot for all your support so far!

**Comment by IzzySoft** [comment]
Works great on the SHIFT6mq as well. As you gave permission, I'd put it online – but the version I have has no matching signature, and carries the same versionName/versionCode as the last one already there. Would you mind a last compile with `versionCode: 11` and `versionName: "1.4.7"` and your sig? I'd then add it to my repo together with a note that it's no longer developed (but works fine at least up to Android 11), and your "may work" warning from above. Thanks in advance!
